[SYCL][ESIMD] Add function that let iterating over types and dims

vasilytric opened the pull request:
The new version of `for_all_types` will be useful in tests for `simd`
This feature lets us avoid using a lots of aliases duplicating instead of this just using  `values_pack`, it can be found [here](https://github.com/intel/llvm-test-suite/pull/588/files#diff-3bbc1ab1445c827282cc08d73f76b58cf553d9ced2c799ba46e6148583896d22R24-R28)

The testing result is the same as before using this feature

vasilytric committed:
[SYCL][ESIMD] Add functions that let iterating over types and dims

Was added function for_types_and_dims and was updated "ctor_default"
test

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -116,44 +116,19 @@ template <typename DataT, int NumElems, typename TestCaseT> struct test {
-  sycl::queue queue{esimd_test::ESIMDSelector{},
-                    esimd_test::createExceptionHandler()};
+  sycl::queue q{esimd_test::ESIMDSelector{},
+                esimd_test::createExceptionHandler()};
```
Why this change needed?

vasilytric replied:
To have a statement, that used `queue` in one string. Now this change was reverted

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -116,44 +116,19 @@ template <typename DataT, int NumElems, typename TestCaseT> struct test {
-  passed &=
-      run_verification_with_chosen_test_type<rval_in_expression>(queue, types);
-  passed &= run_verification_with_chosen_test_type<const_ref>(queue, types);
+  passed &= for_types_and_dims<test, initializer>(types, vector_lengths, q);
```
```suggestion
  passed &= for_types_and_dims<test, initializer>(types, dims, q);
```
Let's use a solid naming over the tests, with no synonyms for the same term. So if you have a `dimensions` term defined by using `for_all_types_and_dims`, let's use it over the code.

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/588/commits/f4c3940c7788727dae45410460dde6655274266d#diff-6d44d6893e30c56cec1c2e698ea43849081d8b092f7a87532bd567294d14ed69R120)

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -86,28 +91,42 @@ template <tested_types required> auto get_tested_types() {
   }
 }
 
+auto get_all_vector_lengths() { return values_pack<1, 8, 16, 32>(); }
```
```suggestion
// Factory method to retrieve pre-defined type_pack with simd vector
// dimensions, to have the same default type coverage over the tests
auto get_tested_dimensions() { return values_pack<1, 8, 16, 32>(); }
```
As a suggestion.
(1) Once again, on uniformity. You already have some `get_tested_types()` factory function. Why make the name for similar functionality so different? It's more hard to follow, it doesn't show itself nearby in case of IDE code auto-complete features usage.
(2) Comment missing

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/588/commits/b359d36a0cf7e324f477dea39bd6f3736b648c55#diff-3bbc1ab1445c827282cc08d73f76b58cf553d9ced2c799ba46e6148583896d22R94-R104)

yuriykoch replied:
What do we need the `enum class tested_dims { all };` for?
I mean, when we need some custom values pack, we might just use it explicitly on a test-wide basis.. Am I missing something?

vasilytric replied:
to have prepared enum for the future, now I don't know which dimensions could be useful
may be will be useful another sets of dimensions for different tests

yuriykoch replied:
Do you mean something like `single` with `values_pack<8>` for example? Let's add it then, so that test would use the same value for case when all tests should be run on some `single` dimension from the full set)

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -86,28 +91,42 @@ template <tested_types required> auto get_tested_types() {
-       ++typeNameIndex,
-       0 // Dummy initialization value
-       )...};
-  // Every initializer clause is sequenced before any initializer clause that
-  // follows it in the braced-init-list. Every expression in comma operator is
-  // also strictly sequenced. So we can use increment safely. We still should
-  // discard dummy results, but this initialization should not be optimized out
-  // due side-effects
```
Are you sure the comment on how it works should be removed?

vasilytric replied:
I've reverted this comment [here](https://github.com/intel/llvm-test-suite/pull/588/commits/b359d36a0cf7e324f477dea39bd6f3736b648c55#diff-3bbc1ab1445c827282cc08d73f76b58cf553d9ced2c799ba46e6148583896d22R120-R124)

yuriykoch replied:
Please, note the comment should be aligned with the changes you've made. There is neither dummy results nor initializer clause in a new code.

vasilytric replied:
> Please, note the comment should be aligned with the changes you've made. There is neither dummy results nor initializer clause in a new code.

Could you suggest some comment to left?

yuriykoch replied:
Sure, I'll make a suggestion then once we handle the question of the expansion context. Thanks for a prompt reply)

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -86,28 +91,42 @@ template <tested_types required> auto get_tested_types() {
+template <template <typename, int, typename...> class Action,
+          typename... ActionArgsT, typename... TypesT, int... Ints,
+          typename... ArgsT>
+inline bool for_types_and_dims(const named_type_pack<TypesT...> &type_list,
```
```suggestion
inline bool for_all_types_and_dims(const named_type_pack<TypesT...> &type_list,
```
Once again, out of uniformity and to make the full set of combinations used explicit.

vasilytric replied:
I've reverted this comment [here](https://github.com/intel/llvm-test-suite/pull/588/commits/b359d36a0cf7e324f477dea39bd6f3736b648c55#diff-3bbc1ab1445c827282cc08d73f76b58cf553d9ced2c799ba46e6148583896d22R133)

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -86,28 +91,42 @@ template <tested_types required> auto get_tested_types() {
+
+// Calls for_all_types for each vector length by values_pack instance
+template <template <typename, int, typename...> class Action,
+          typename... ActionArgsT, typename... TypesT, int... Ints,
```
```suggestion
          typename... ActionArgsT, typename... TypesT, int... Dims,
```
That's dimensions mentioned in the function name, right?

vasilytric replied:
I've reverted this comment [here](https://github.com/intel/llvm-test-suite/pull/588/commits/b359d36a0cf7e324f477dea39bd6f3736b648c55#diff-3bbc1ab1445c827282cc08d73f76b58cf553d9ced2c799ba46e6148583896d22R130-R136)

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -86,28 +91,42 @@ template <tested_types required> auto get_tested_types() {
-inline bool for_all_types(const named_type_pack<types...> &typeList,
-                          argsT &&... args) {
+template <template <typename, int, typename...> class Action, int N,
+          typename... ActionArgsT, typename... TypesT, typename... ArgsT>
```
```suggestion
          typename... ActionArgsT, typename... Types, typename... ArgsT>
```
Minor: that's more like a pack of types, so it might be not needed to additionally state that it's a "type". What do you think?

vasilytric replied:
I've reverted this comment [here](https://github.com/intel/llvm-test-suite/pull/588/commits/b359d36a0cf7e324f477dea39bd6f3736b648c55#diff-3bbc1ab1445c827282cc08d73f76b58cf553d9ced2c799ba46e6148583896d22R130-R136)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -42,7 +42,7 @@ struct initializer {
 // Descriptor class for the case of calling constructor in variable declaration
 // context
-struct var_declaration {
+struct var_dec {
```
```suggestion
struct var_decl {
```
The `dec` usually stands for the `decimal`

vasilytric replied:
I've reverted this comment [here](https://github.com/intel/llvm-test-suite/pull/588/commits/f4c3940c7788727dae45410460dde6655274266d#diff-e20a89556dfa9ec98f8eec4bea698de3975b780ff24a3bde1ced890af1d16581L46-L47)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -54,7 +54,7 @@ struct var_declaration {
 // Descriptor class for the case of calling constructor in rvalue in an
 // expression context
-struct rval_in_expression {
+struct rval_in_express {
```
```suggestion
struct rval_in_expr {
```

vasilytric replied:
I've reverted this comment [here](https://github.com/intel/llvm-test-suite/pull/588/commits/f4c3940c7788727dae45410460dde6655274266d#diff-e20a89556dfa9ec98f8eec4bea698de3975b780ff24a3bde1ced890af1d16581L60-R61)

yuriykoch reviewed (commented):
Would you mind to also update the `ctor_copy.cpp` as it's merged already?

vasilytric committed:
Merge branch 'intel' into iterate_over_types_and_dims

vasilytric committed:
[SYCL][ESIMD] Small updates in ctor_copy.cpp and ctor_default.cpp

- New struct names with lesser character names provides write our
  statements in one string
- Now ctor_copy.cpp used "for_all_types_and_dims" function

vasilytric committed:
[SYCL][ESIMD] Updates in type_coverage.hpp

- Add factory method get_tested_dimensions, so we have same logic with
  retreiving tested types
- Reverted comment that how "for_all_types" works

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -86,28 +91,62 @@ template <tested_types required> auto get_tested_types() {
-       )...};
+  size_t type_name_index = 0;
+
+  ((passed &= Action<Types, N, ActionArgsT...>{}(
+        std::forward<ArgsT>(args)..., type_list.names[type_name_index]),
+    ++type_name_index),
```
```suggestion
  (passed &= Action<Types, N, ActionArgsT...>{}(
        std::forward<ArgsT>(args)..., type_list.names[type_name_index]),
    ++type_name_index,
```
Please correct me if I'm wrong, but wouldn't it give the same result due to strict sequencing of every expression delimited by comma operator?
And one more thing that we need to check.. I'm not able to map the expansion pack context to the list of the allowed ones: see "expansion loci" in https://en.cppreference.com/w/cpp/language/parameter_pack. Looks like we need to add a comment stating the expansion pack context explicitly, just to be aware of possible problems on code update.

vasilytric replied:
when I try to apply this comment I've obtained this compilation error:
```
SYCL\ESIMD\api\functional\ctors/../type_coverage.hpp:119:4: error: expected ')'
   ...);
   ^
SYCL\ESIMD\api\functional\ctors/../type_coverage.hpp:116:3: note: to match this '('
  (passed &= Action<Types, N, ActionArgsT...>{}(
  ^
SYCL\ESIMD\api\functional\ctors/../type_coverage.hpp:116:4: error: expression contains unexpanded parameter pack 'Types'
  (passed &= Action<Types, N, ActionArgsT...>{}(
   ^                ~~~~~
SYCL\ESIMD\api\functional\ctors\ctor_default.cpp:128:13: note: in instantiation of function template specialization 'esimd_test::api::functional::for_all_types_and_dims<test, initializer, char, unsigned char, signed char, short, unsigned
      short, int, unsigned int, long, unsigned long, float, double, long long, unsigned long long, 1, 8, 16, 32, cl::sycl::queue &>' requested here
  passed &= for_all_types_and_dims<test, initializer>(types, dims, queue);
            ^
1 warning and 2 errors generated.
```

I think that final simplification of this solution due to I tried many different simplifications of this function and only this version works correctly and without compiled without errors

yuriykoch replied:
OK, I see now, that's an unary right fold expression available since C++17.
Therefore something like:
```
  const auto types = named_type_pack<foo, boo, bar>("foo", "boo", "bar");
  for_all_types<Action, 1>(types);
```
will result in following `named_type_pack`:
```
  named_type_pack::names[3] == {"foo", "boo", "bar"};
```
and in the following unfolded expression:
```
  size_t type_name_index = 0;

  ((passed &= Action<foo, 1>{}("foo"),
    type_name_index = 1),
    (
      (passed &= Action<boo, 1>{}("boo"),
       type_name_index = 2),
      (
        (passed &= Action<bar, 1>{}("bar"),
         type_name_index = 3),
      )
    )
  )
```

And therefore brackets are required due to lower precedence value of the increment operator.
Am I right?

yuriykoch replied:
I would suggest the following comment then:
```
  // The unary right fold expression is used for parameter pack expansion.
  // Every expression with comma operator is strictly sequenced, so we can 
  // increment safely. And of course the fold expression would not be optimized
  // out due to side-effects.
  // Additional pair of brackets is required because of precedence of increment
  // operator relative to the comma operator.
  //
  // Note that there is actually no difference in left or right fold expression
  // for the comma operator, as it would give the same order of actions
  // execution and the same order of the type name index increment: both the
  // "(expr0, (exr1, expr2))" and "((expr0, expr1), expr2)" would give the same
  //  result as simple "expr0, expr1, expr2"
```
Thanks for the great idea!

vasilytric replied:
yes, order is right, You is right.

vasilytric replied:
The comments was replaced in this commit: 9ce01972c9d4b14c835faf126b6913a107449067

vasilytric committed:
[SYCL][ESIMD] Rename "get_tested_dimensions" to "get_all_dimensions"

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -86,28 +91,62 @@ template <tested_types required> auto get_tested_types() {
+          typename... ActionArgsT, typename... Types, int... Dims,
+          typename... ArgsT>
+inline bool for_all_types_and_dims(const named_type_pack<Types...> &type_list,
+                                   const values_pack<Dims...> &intList,
```
```suggestion
                                   const values_pack<Dims...> &dim_list,
```

vasilytric replied:
I've applied this comment in this commit: 9ce01972c9d4b14c835faf126b6913a107449067

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -86,28 +91,62 @@ template <tested_types required> auto get_tested_types() {
+                                   ArgsT &&... args) {
+  bool passed{true};
+
+  // run action for each value from values_pack... parameter pack
+  int int_index = 0;
+
+  ((passed &= for_all_types<Action, Dims, ActionArgsT...>(
+        type_list, std::forward<ArgsT>(args)...),
+    ++int_index),
+   ...);
   // Every initializer clause is sequenced before any initializer clause that
   // follows it in the braced-init-list. Every expression in comma operator is
   // also strictly sequenced. So we can use increment safely. We still should
   // discard dummy results, but this initialization should not be optimized out
   // due side-effects
```
```suggestion
  // Run action for each value from values_pack... parameter pack
  ((passed &= for_all_types<Action, Dims, ActionArgsT...>(
        type_list, std::forward<ArgsT>(args)...)),
   ...);
  // The unary right fold expression is used for parameter pack expansion.
  // An additional pair of brackets is required because of precedence of any
  // operator relatively to the comma operator.
```

vasilytric replied:
The comments was replaced in this commit: 9ce01972c9d4b14c835faf126b6913a107449067

vasilytric committed:
[SYCL][ESIMD] Changes in ESIMD/api/functional/type_coverage.hpp

- Updated comments
- Removed unnecessary variable

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

vasilytric commented:
@sndmitriev Could you review this PR?

sndmitriev reviewed (approved):
LGTM

vasilytric commented:
@sndmitriev I am not authorized to merge this PR, could you merge it?

sndmitriev commented:
> @sndmitriev I am not authorized to merge this PR, could you merge it?

Neither do I. @kbobrovs, can you please merge this PR?

vasilytric commented:
@v-klochkov @sndmitriev thanks Vyacheslav and Serguei
